program/rust/src/tests/test_upd_product.rs

```diff
@@ -0,0 +1,163 @@
+use std::mem::size_of;
+
+use solana_program::account_info::AccountInfo;
+use solana_program::clock::Epoch;
+use solana_program::native_token::LAMPORTS_PER_SOL;
+use solana_program::program_error::ProgramError;
+use solana_program::pubkey::Pubkey;
+use solana_program::rent::Rent;
+use solana_program::system_program;
+
+use crate::c_oracle_header::{
+    cmd_hdr_t,
+    cmd_upd_product_t,
+    command_t_e_cmd_upd_product,
+    pc_prod_t,
+    PC_PROD_ACC_SIZE,
+    PC_VERSION,
+};
+use crate::deserialize::load_mut;
+use crate::rust_oracle::{
+    initialize_product_account,
+    load_product_account_mut,
+    read_pc_str_t,
+    try_convert,
+    upd_product,
+};
+
+#[test]
+fn test_upd_product() {
+    let mut instruction_data = [0u8; PC_PROD_ACC_SIZE as usize];
+
+    let program_id = Pubkey::new_unique();
+    let funding_key = Pubkey::new_unique();
+    let product_key = Pubkey::new_unique();
+
+    let system_program = system_program::id();
+    let mut funding_balance = LAMPORTS_PER_SOL.clone();
+    let funding_account = AccountInfo::new(
+        &funding_key,
+        true,
+        true,
+        &mut funding_balance,
+        &mut [],
+        &system_program,
+        false,
+        Epoch::default(),
+    );
+
+    let mut product_balance = Rent::minimum_balance(&Rent::default(), PC_PROD_ACC_SIZE as usize);
+    let mut prod_raw_data = [0u8; PC_PROD_ACC_SIZE as usize];
+    let product_account = AccountInfo::new(
+        &product_key,
+        true,
+        true,
+        &mut product_balance,
+        &mut prod_raw_data,
+        &program_id,
+        false,
+        Epoch::default(),
+    );
+
+    initialize_product_account(&product_account, PC_VERSION).unwrap();
+
+    let kvs = ["foo", "barz"];
+    let size = populate_instruction(&mut instruction_data, &kvs);
+    assert!(upd_product(
+        &program_id,
+        &[funding_account.clone(), product_account.clone()],
+        &instruction_data[..size]
+    )
+    .is_ok());
+    assert!(account_has_key_values(&product_account, &kvs).unwrap());
```

Comment:
`.is_ok()` instead of unwrap?

Reply:
Ok this is fine, you're actually checking the bool. I would still use `unwrap_or`